File: onstock/internal/store/suppliers.go
package store

import (
	"database/sql"
	"errors"
	"strings"
)

func (s *Store) ListSuppliers(includeInactive bool) ([]Supplier, error) {
	q := `SELECT id, name, rtn, contact_name, phone, email, address, notes, active, created_at FROM suppliers`
	if !includeInactive {
		q += ` WHERE active=1`
	}
	q += ` ORDER BY name`
	rows, err := s.db.Query(q)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []Supplier{}
	for rows.Next() {
		var sp Supplier
		var active int
		if err := rows.Scan(&sp.ID, &sp.Name, &sp.RTN, &sp.ContactName, &sp.Phone, &sp.Email,
			&sp.Address, &sp.Notes, &active, &sp.CreatedAt); err != nil {
			return nil, err
		}
		sp.Active = active == 1
		out = append(out, sp)
	}
	return out, rows.Err()
}

func (s *Store) GetSupplier(id int64) (Supplier, error) {
	var sp Supplier
	var active int
	err := s.db.QueryRow(`SELECT id, name, rtn, contact_name, phone, email, address, notes, active, created_at
	  FROM suppliers WHERE id=?`, id).
		Scan(&sp.ID, &sp.Name, &sp.RTN, &sp.ContactName, &sp.Phone, &sp.Email, &sp.Address, &sp.Notes, &active, &sp.CreatedAt)
	if errors.Is(err, sql.ErrNoRows) {
		return sp, ErrNotFound
	}
	sp.Active = active == 1
	return sp, err
}

func (s *Store) CreateSupplier(sp Supplier) (Supplier, error) {
	if strings.TrimSpace(sp.Name) == "" {
		return sp, errors.New("el nombre es obligatorio")
	}
	res, err := s.db.Exec(`INSERT INTO suppliers (name, rtn, contact_name, phone, email, address, notes, active)
	  VALUES (?,?,?,?,?,?,?,1)`,
		strings.TrimSpace(sp.Name), sp.RTN, sp.ContactName, sp.Phone, sp.Email, sp.Address, sp.Notes)
	if err != nil {
		return sp, err
	}
	id, _ := res.LastInsertId()
	return s.GetSupplier(id)
}

func (s *Store) UpdateSupplier(id int64, sp Supplier) (Supplier, error) {
	if strings.TrimSpace(sp.Name) == "" {
		return sp, errors.New("el nombre es obligatorio")
	}
	res, err := s.db.Exec(`UPDATE suppliers SET name=?, rtn=?, contact_name=?, phone=?, email=?, address=?, notes=?, active=?
	  WHERE id=?`,
		strings.TrimSpace(sp.Name), sp.RTN, sp.ContactName, sp.Phone, sp.Email, sp.Address, sp.Notes, boolToInt(sp.Active), id)
	if err != nil {
		return sp, err
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return sp, ErrNotFound
	}
	return s.GetSupplier(id)
}

// DeleteSupplier desactiva si tiene historial (productos, órdenes o gastos); elimina si no.
func (s *Store) DeleteSupplier(id int64) error {
	var refs int
	err := s.db.QueryRow(`SELECT
	  (SELECT COUNT(*) FROM products WHERE supplier_id=?) +
	  (SELECT COUNT(*) FROM purchase_orders WHERE supplier_id=?) +
	  (SELECT COUNT(*) FROM expenses WHERE supplier_id=?)`, id, id, id).Scan(&refs)
	if err != nil {
		return err
	}
	if refs > 0 {
		_, err = s.db.Exec(`UPDATE suppliers SET active=0 WHERE id=?`, id)
		return err
	}
	_, err = s.db.Exec(`DELETE FROM suppliers WHERE id=?`, id)
	return err
}

File: onstock/server/store/suppliers.js
// Traducción de internal/store/suppliers.go.
import { ErrNotFound, BizError } from '../db.js';

function scanSupplier(sp) {
  return {
    id: sp.id,
    name: sp.name,
    rtn: sp.rtn,
    contact_name: sp.contact_name,
    phone: sp.phone,
    email: sp.email,
    address: sp.address,
    notes: sp.notes,
    active: sp.active === 1,
    created_at: sp.created_at,
  };
}

export function listSuppliers(db, includeInactive) {
  let q = 'SELECT id, name, rtn, contact_name, phone, email, address, notes, active, created_at FROM suppliers';
  if (!includeInactive) q += ' WHERE active=1';
  q += ' ORDER BY name';
  return db.all(q).map(scanSupplier);
}

export function getSupplier(db, id) {
  const row = db.get(
    `SELECT id, name, rtn, contact_name, phone, email, address, notes, active, created_at
	  FROM suppliers WHERE id=?`, id,
  );
  if (!row) throw new ErrNotFound();
  return scanSupplier(row);
}

export function createSupplier(db, sp) {
  if (!String(sp.name ?? '').trim()) throw new BizError('el nombre es obligatorio');
  const res = db.run(
    `INSERT INTO suppliers (name, rtn, contact_name, phone, email, address, notes, active)
	  VALUES (?,?,?,?,?,?,?,1)`,
    String(sp.name).trim(), sp.rtn ?? '', sp.contact_name ?? '', sp.phone ?? '',
    sp.email ?? '', sp.address ?? '', sp.notes ?? '',
  );
  return getSupplier(db, res.lastInsertRowid);
}

export function updateSupplier(db, id, sp) {
  if (!String(sp.name ?? '').trim()) throw new BizError('el nombre es obligatorio');
  const res = db.run(
    `UPDATE suppliers SET name=?, rtn=?, contact_name=?, phone=?, email=?, address=?, notes=?, active=?
	  WHERE id=?`,
    String(sp.name).trim(), sp.rtn ?? '', sp.contact_name ?? '', sp.phone ?? '',
    sp.email ?? '', sp.address ?? '', sp.notes ?? '', sp.active ? 1 : 0, id,
  );
  if (res.changes === 0) throw new ErrNotFound();
  return getSupplier(db, id);
}

// deleteSupplier desactiva si tiene historial (productos, órdenes o gastos); elimina si no.
export function deleteSupplier(db, id) {
  const refs = db.scalar(
    `SELECT
	  (SELECT COUNT(*) FROM products WHERE supplier_id=?) +
	  (SELECT COUNT(*) FROM purchase_orders WHERE supplier_id=?) +
	  (SELECT COUNT(*) FROM expenses WHERE supplier_id=?)`,
    id, id, id,
  );
  if (refs > 0) {
    db.run('UPDATE suppliers SET active=0 WHERE id=?', id);
    return;
  }
  db.run('DELETE FROM suppliers WHERE id=?', id);
}
